### ISSUE TITLE
Scripted events in conversation happen too soon

### ISSUE THREAD

**Issue by Nindaleth** [Try/OpenGothic#289]
By "scripted events" I mean things such as "New Diary Entry", "Experience + 100", "Item received" etc. that happen while talking to a character.

One simple reproduction is with Cavalorn right at the start. After you clear out the bandit cave, he talks to you. In vanilla Gothic you get a new diary entry and some experience immediately, then talk a little, he says "here's a couple of coins", then you also get some gold. In OpenGothic you immediately get the diary entry, experience and also the gold the moment Cavalorn starts talking to you.

Another easy repro is when making a donation to Daron (the priest of Innos). In vanilla, he looks through your money pouch, then states the amount to donate, then that gold is taken. In OpenGothic, you lose the gold the moment you choose the donating conversation option.

**Comment by Try** [comment]
Hi, @Nindaleth !

I'm aware, but it's not quite clear what is suppose correct behavior. Here is Cavalor scrip for instance:
```c++

func void DIA_Addon_Cavalorn_Beutel_Info()
{
	....
	AI_Output(self,other,"DIA_Addon_Cavalorn_Beutel_08_07");	//Do you have it with you? I'll give you 100 gold coins for it.
	TOPIC_End_CavalornTheHut = TRUE;
	B_GivePlayerXP(XP_Ambient);
	...
};
```
`AI_Output` - is queue operation, that doesn't executes immediately.
`B_GivePlayerXP` - is helper function, that makes call to:
```
PrintScreen(concatText,-1,YPOS_XPGained,FONT_ScreenSmall,2);
```
And `PrintScreen` is suppose to be immediate function.

Most likely assumption about "immediate function" is not 100% correct, but then it's not clear how it should be executed.

**Comment by Nindaleth** [comment]
I think in most cases the XP and diary entry actually are immediate (right after starting the conversation or picking a dialog option), but item gain/loss only happens after a certain line a few dialog boxes later.

I'll record a few comparison videos if I have time in the following days; maybe the vanilla video, OpenGothic video and script side-by-side help to discover some exceptions to the immediate behaviour.

**Comment by Nindaleth** [comment]
Here are the comparison videos. It's apparent that diary entries and XP gains are always immediate, but the item/gold gains/losses are delayed in original and immediate in OpenGothic.

All the videos start with the original, followed by OpenGothic. They are tightly cut to fit the Github attachment size limit, also the 2nd half of each video is empty and can be ignored; it's some issue introduced by my encoding..

### Getting gold from Cavalorn

https://user-images.githubusercontent.com/6383928/189454723-540466e7-82cc-462d-a3d4-e410fc429476.mp4

### Giving gold to Daron

https://user-images.githubusercontent.com/6383928/189454741-d87222ab-84a3-4101-bb31-62e21c06bea2.mp4

### Getting wine from Maleth

https://user-images.githubusercontent.com/6383928/189454941-78a9358e-4250-41c1-8df2-91427714610f.mp4

### Giving wine to Vino

https://user-images.githubusercontent.com/6383928/189454963-d7978b16-f3ae-4e7f-be1f-6f72aebb78f5.mp4

**Comment by Nindaleth** [comment]
There's one additional excellent point to reproduce this - selling hides to Bosper. In the original, every hide type is sold separately (Bosper comments it and individual "x items given", "x gold received" print is shown). In OpenGothic, all of the separate giving and receiving item print happens all at once, resulting in unreadable mishmash of text displayed on screen at the beginning of the dialog option.

**Comment by Try** [comment]
> It's apparent that diary entries and XP gains are always immediate, but the item/gold gains/losses are delayed in original and immediate in OpenGothic.

Yes, and that a big problem. Any of text-messages are `printscreen`-like commands and from api point of view I can't see a reason why the should be executed differently :(

**Comment by Nindaleth** [comment]
Maybe there's a condition parameter hidden within the `printscreen` function or something like that, like that `2` in your example? Is there a simple way I can parse these scripts on my own?

**Comment by Try** [comment]
@Nindaleth scripts are there on github:
https://github.com/IDizor/GothicScripts/blob/master/SolGothicScriptsOriginal/Prjgothic/Story/Startup.d

Also there are a few decopiler-kind of projects, that I haven't tested myself:
https://github.com/auronen/DecDat/releases/tag/v1.0b
gothic sourcer

**Comment by Nindaleth** [comment]
Thanks @Try! I agree it seems that the `AI_PrintScreen` is always immediate. But I can see another pattern there. Going by the first video (killing Brago and two other bandits with Cavalorn, then talking to him).

It starts at `DIA_Addon_Cavalorn_BragoKilled_Info` ("So, that's done. Hah. They shouldn't have messed with me.") and it just calls a different function `B_Addon_Cavalorn_VatrasBrief` which gives XP and adds a journal entry, the last spoken sentence in that called function is "Otherwise, you may be taken for a bandit. Here's a couple of coins." As is obvious from the video, everything within these two functions is processed immediately - the XP and a journal entry are given right at the start. 

But we can also see that the `CreateInvItems` and `B_GiveInvItems` functions are evaluated (or, at least their screen printing is) only after the respective dialog line in `AI_Output` is finished. As if the script was pre-evaluated in full and then the item creating/giving function calls were postponed to be in sync with the dialog.

**Comment by Try** [comment]
Oh, wait a momoent:
```c++
void Gothic::ai_printscreen(Daedalus::DaedalusVM& vm) {
  // TODO: print-screen queue
  ...
```
Sorry, I forgot to implement it ;D

**Comment by Try** [comment]
In modkit docs `AI_PrintScreen` is not clearly defined: it doesn't take an `npc` argument.

So far enqueuing to `self` feels as correct most option.

**Comment by Try** [comment]
Tested on Maleth, Vino and Daron - work correct now
